core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -905,33 +893,26 @@ class TimeSeriesShard(val dataset: Dataset,
 
   private[memstore] val addPartitionsDisabled = AtomicBoolean(false)
 
-  //scalastyle:off null
-  private[filodb] def getOrAddPartition(recordBase: Any, recordOff: Long, group: Int, ingestOffset: Long)
-      : FiloPartition = {
-    var stamp = partSetLock.tryOptimisticRead()
-    if (stamp != 0) {
-      val part = partSet.getWithIngestBR(recordBase, recordOff)
-      if (part != null && partSetLock.validate(stamp)) {
-        return part
-      }
-    }
-
-    stamp = partSetLock.readLock()
-    try {
-      val part = partSet.getWithIngestBR(recordBase, recordOff)
-      if (part != null) {
-        return part
-      }
-    } finally {
-      partSetLock.unlockRead(stamp)
+  // scalastyle:off null
+  private[filodb] def getOrAddPartition(recordBase: Any, recordOff: Long, group: Int, ingestOffset: Long) = {
+    var part = partSet.getWithIngestBR(recordBase, recordOff)
+    if (part == null) {
+      part = addPartition(recordBase, recordOff, group, ingestOffset)
     }
+    part
+  }
 
-    stamp = partSetLock.writeLock()
-    var part = partSet.getWithIngestBR(recordBase, recordOff)
-    var newPart: TimeSeriesPartition = null
+  private def addPartition(recordBase: Any, recordOff: Long, group: Int, ingestOffset: Long) = {
+    val stamp = partSetLock.writeLock()
 
-    try {
-      if (part == null) {
+    // Double check just in case multiple threads are calling this method.
+    val part = partSet.getWithIngestBR(recordBase, recordOff)
+    if (part != null) {
```

Comment:
@broneill I'll trust your intuition that the double check takes care of diff race conditions.   :)